h/api/models/document.py

```diff
@@ -181,6 +184,148 @@ def __repr__(self):
+
+    :param uri: the .uri property of the DocumentURI
+    :type uri: unicode
+
+    :param type: the .type property of the DocumentURI
+    :type type: unicode
+
+    :param content_type: the .content_type property of the DocumentURI
+    :type content_type: unicode
+
+    :param document: the Document that the new DocumentURI will belong to, if a
+        new DocumentURI is created
+    :type document: h.api.models.Document
+
+    :param created: the time that will be used as the .created time for the new
+        DocumentURI, if a new one is created
+    :type created: datetime.datetime
+
+    :param updated: the time that will be set as the .updated time for the new
+        or existing DocumentURI
+    :type updated: datetime.datetime
+
+    """
+    docuri = DocumentURI.query.filter(
+        DocumentURI.claimant_normalized == uri_normalize(claimant),
+        DocumentURI.uri_normalized == uri_normalize(uri),
+        DocumentURI.type == type,
+        DocumentURI.content_type == content_type).first()
+
+    if docuri is None:
+        docuri = DocumentURI(claimant=claimant,
+                             uri=uri,
+                             type=type,
+                             content_type=content_type,
+                             document=document,
+                             created=created,
+                             updated=updated)
+        session.add(docuri)
+    elif not docuri.document == document:
+        log.warn("Found DocumentURI (id: %d)'s document_id (%d) doesn't match "
+                 "given Document's id (%d)",
+                 docuri.id, docuri.document_id, document.id)
+
+    docuri.updated = updated
+
+
+def create_or_update_document_meta(session,
+                                   claimant,
+                                   type,
+                                   value,
+                                   document,
+                                   created,
+                                   updated):
+    """
+    Create or update a DocumentMeta with the given parameters.
+
+    If an equivalent DocumentMeta already exists in the database then its value
+    and updated time will be updated.
+
+    If no equivalent DocumentMeta exists in the database then a new one will be
+    created and added to the database.
+
+    To be considered "equivalent" an existing DocumentMeta must have the given
+    claimant and type, but its value, document and created and updated times
+    needn't match the given ones.
+
+    :param session: the database session
+    :type session: sqlalchemy.orm.session.Session
+
+    :param claimant: the value to use for the DocumentMeta's claimant attribute
+        if a new DocumentMeta is created
+    :type claimant: unicode
+
+    :param type: the value of the new or existing DocumentMeta's type attribute
+    :type type: unicode
+
+    :param value: the value to set the new or existing DocumentMeta's value
+        attribute to
+    :type value: unicode
```

Comment:
Which actually is quite annoying because when you pass it a string/unicode SQLAlchemy/Postgres will just write the string as a list of individual characters :disappointed: